### PR TITLE
octree uses shared pointers

### PR DESCRIPTION
**Pull request by sethalves**
- Octree and EntityTree us shared pointers

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31fcd55d473ce7a6c9008a373c79b2ca67e3c934 available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8a60bba4eba841ae551f811a4b7939f69a785146 available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9c4fdfd65351d9b33d5615c953100480db8a29ef available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: babe6179c31d072b77f4b75b23b1249fab6546a2 available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

File: assignment-client/src/octree/OctreeQueryNode.cpp [hunk 1]
```diff
@@ -338,9 +338,9 @@ void OctreeQueryNode::dumpOutOfView() {
     int outOfView = 0;
     OctreeElementBag tempBag;
     while (!elementBag.isEmpty()) {
-        OctreeElement* node = elementBag.extract();
-        if (node->isInView(_currentViewFrustum)) {
-            tempBag.insert(node);
+        OctreeElementPointer elt = elementBag.extract();
```
Comment (ZappoMan) [inline assignment-client/src/octree/OctreeQueryNode.cpp]: coding standard: elt is not a clear variable name

File: assignment-client/src/octree/OctreeQueryNode.cpp [hunk 2]
```diff
@@ -348,9 +348,9 @@ void OctreeQueryNode::dumpOutOfView() {
     }
     if (stillInView > 0) {
         while (!tempBag.isEmpty()) {
-            OctreeElement* node = tempBag.extract();
-            if (node->isInView(_currentViewFrustum)) {
-                elementBag.insert(node);
+            OctreeElementPointer elt = tempBag.extract();
```
Comment (ZappoMan) [inline assignment-client/src/octree/OctreeQueryNode.cpp]: coding standard: elt is not a clear variable name

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7b656d5eade4509b3cef3ab4486522743752bc81 available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.h [hunk 3]
```diff
@@ -123,15 +123,19 @@ public slots:
     void setDisplayModelBounds(bool value) { _displayModelBounds = value; }
     void setDisplayModelElementProxy(bool value) { _displayModelElementProxy = value; }
     void setDontDoPrecisionPicking(bool value) { _dontDoPrecisionPicking = value; }
-    
+
 protected:
-    virtual Octree* createTree() { return new EntityTree(true); }
+    virtual OctreePointer createTree() {
+        EntityTreePointer newTree = EntityTreePointer(new EntityTree(true));
+        newTree->createRootElement();
```
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: Why do you need this? Doesn't the EntityTree ctr correctly initialize it's root element? In the production code it does...

File: libraries/entities/src/EntityTree.cpp [hunk 4]
```diff
@@ -31,7 +31,6 @@ EntityTree::EntityTree(bool shouldReaverage) :
     _fbxService(NULL),
     _simulation(NULL)
 {
-    _rootElement = createNewElement();
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTree.cpp]: why did you remove this??

File: libraries/entities/src/EntityTree.cpp [hunk 5]
```diff
@@ -39,10 +38,19 @@ EntityTree::~EntityTree() {
...
-    newElement->setTree(this);
-    return newElement;
+void EntityTree::createRootElement() {
+    _rootElement = createNewElement();
+}
+
+OctreeElementPointer EntityTree::createNewElement(unsigned char* octalCode) {
+    EntityTreeElementPointer newElement = EntityTreeElementPointer(new EntityTreeElement(octalCode),
+                                                                   // see comment int EntityTreeElement::createNewElement
+                                                                   [=](EntityTreeElement* elt) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTree.cpp]: coding standard: variable names should be descriptive

File: libraries/entities/src/EntityTreeElement.cpp [hunk 6]
```diff
@@ -30,26 +30,46 @@ EntityTreeElement::~EntityTreeElement() {
...
+        EntityTreeElementPointer(new EntityTreeElement(octalCode),
+                                 // This is a little bit horrible, but I haven't found a better way.  The OctreeElement
+                                 // destructor used to call notifyDeleteHooks(), which calls zero or more of
+                                 //   OctreeElementDeleteHook::elementDeleted
+                                 // which (now) expects an OctreeElementPointer argument.  The destructor doesn't have
+                                 // access to the shared pointer (which has had its reference count drop to zero,
+                                 // or the destructor wouldn't have been called).  The destructor also can't
+                                 // make a new shared pointer -- shared_from_this() is forbidden in a destructor, and
+                                 // using OctreeElementPointer(this) also fails.  So, I've installed a custom deleter:
+                                 [=](EntityTreeElement* elt) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTreeElement.cpp]: coding standard: variable names should be descriptive

File: libraries/entities/src/EntityItem.cpp [hunk 7]
```diff
@@ -985,6 +985,12 @@ bool EntityItem::isMoving() const {
     return hasVelocity() || hasAngularVelocity();
 }
 
+EntityTreePointer EntityItem::getTree() const {
+    EntityTreeElementPointer elt = getElement();
```
Comment (ZappoMan) [inline libraries/entities/src/EntityItem.cpp]: coding standard: variable names should be descriptive

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 8]
```diff
@@ -47,19 +47,19 @@ bool EntityScriptingInterface::canRez() {
     return nodeList->getThisNodeCanRez();
 }
 
-void EntityScriptingInterface::setEntityTree(EntityTree* modelTree) {
+void EntityScriptingInterface::setEntityTree(EntityTreePointer modelTree) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: let's fix this old variable name... modelTree => elementTree

File: libraries/entities/src/EntityTree.cpp [hunk 9]
```diff
@@ -39,10 +38,19 @@ EntityTree::~EntityTree() {
...
+                                                                       EntityTreeElementPointer tmpSharedPointer(elt);
+                                                                       elt->notifyDeleteHooks();
+                                                                   });
+    newElement->setTree(std::static_pointer_cast<EntityTree>(shared_from_this()));
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTree.cpp]: I'm confused by this code and the other places where you create EntityTreePointers... specifically... 

```
EntityTreePointer tree = EntityTreePointer(new EntityTree(true));
tree->createRootElement();
```
Maybe I've misunderstood it... but I believe the shared pointer you get from `shared_from_this()` will be a different pointer (with different refcount) than any pointer you get from ` EntityTreePointer(new EntityTree(true));` so I believe you're going to get mismatched refcounts here.

It seems like you'd want to do something like....

```
EntityTree::getMyEntityTreePointer() { return shared_from_this(); }

virtual OctreePointer createTree() {
    EntityTree* newTree = new EntityTree(true);
    return newTree->getMyEntityTreePointer();
}
```

File: libraries/entities-renderer/src/EntityTreeRenderer.h (hunk 3 above)
Comment (sethalves) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: The problem is that the EntityTreeElement has a backpointer to the Tree -- you can't use shared_from_this() until  something else has taken a shared pointer of "this", meaning you can't use it in the constructor.  So, I have to add the root element after the EntityTree constructor is finished.

File: libraries/entities/src/EntityTree.cpp (hunk 4 above)
Comment (sethalves) [inline libraries/entities/src/EntityTree.cpp]: Same problem here.

File: libraries/entities/src/EntityTree.cpp (hunk 9 above)
Comment (sethalves) [inline libraries/entities/src/EntityTree.cpp]: shared_from_this finds the existing shared pointer for the object.  This is why you can't use it until something already has a shared pointer to "this".

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1c5a04531a69d78a336d3953c96844ccd1aa2ef available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg

**Comment by sethalves** [comment]
test this please

File: libraries/entities-renderer/src/EntityTreeRenderer.h (hunk 3 above)
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.h]: shared pointers SUCK! ;)

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1c5a04531a69d78a336d3953c96844ccd1aa2ef available at: https://gist.github.com/2a19b22c52debb019061<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5734.dmg
